### PR TITLE
Add Linux/Unix support

### PR DESCRIPTION
**Pull request by TheComputerGuy96**
This is currently a draft until I can get SilentPatch actually compiling/working with MinGW GCC (I got the ddraw component to compile though)

### CODE REVIEW

**Comment by CookiePLMonster** [comment]
This looks very promising, looking forward to that!

**Comment by TheComputerGuy96** [comment]
This PR should be okay now (because I tested it with SilentPatchIII/VC that's locally compiled with MinGW GCC and both of those libraries compile/work fine)

**Comment by CookiePLMonster** [comment]
> This PR should be okay now (because I tested it with SilentPatchIII/VC that's locally compiled with MinGW GCC and both of those libraries compile/work fine)

Good stuff! I'll check on Windows locally with SP for II/VC (using a VS2017 toolset) and with any other patch using VS2022, if everything works good then I'll merge. I haven't spotted anything wrong in the review though so it should be all good.

**Comment by CookiePLMonster** [comment]
I see you're still making changes, is this ready to go or should I wait?

**Comment by TheComputerGuy96** [comment]
> I see you're still making changes, is this ready to go or should I wait?

The III/VC parts seem pretty stable (SA has weird segfault issues though) so this is good enough

File: MemoryMgr.GTA.h [hunk 1]
```diff
@@ -20,7 +20,7 @@ namespace Memory
 		ptrdiff_t offset;
 	};
 
-	using AddrVariant = std::variant<uintptr_t, PatternAndOffset>;
+	using AddrVariant = std::variant<intptr_t, uintptr_t, PatternAndOffset>;
```
Comment (CookiePLMonster) [inline MemoryMgr.GTA.h]: I wonder if `uintptr_t` is needed here to begin with. If it's not, it'd simplify the `get_if` accesses quite a lot.

File: MemoryMgr.GTA.h [hunk 2]
```diff
@@ -268,7 +268,9 @@ namespace Memory
 				if ( auto pao = std::get_if<PatternAndOffset>(&address10) ) return HandlePattern( *pao );
 				else
 				{
-					const uintptr_t addr = *std::get_if<uintptr_t>(&address10);
+					const uintptr_t addr = (std::get_if<uintptr_t>(&address10) != nullptr)
+											? *std::get_if<uintptr_t>(&address10)
+											: std::make_unsigned_t<intptr_t>(*std::get_if<intptr_t>(&address10));
```
Comment (CookiePLMonster) [inline MemoryMgr.GTA.h]: That might require `<type_traits>` to be included I think.

File: MemoryMgr.GTA.h (hunk 1 above)
Comment (TheComputerGuy96) [reply MemoryMgr.GTA.h]: I get errors like this without allowing `intptr_t` in AddrVariant:
```
/mnt/storagebeast/Development/General/SilentPatch/SilentPatchSA/AudioHardwareSA.cpp: At global scope:
/mnt/storagebeast/Development/General/SilentPatch/SilentPatchSA/AudioHardwareSA.cpp:6:65: error: no matching function for call to ‘AddressByVersion<void*>(int, int, int)’
    6 | static void* CAEDataStream__Initialise = AddressByVersion<void*>(0x4DC2B0, 0x4DC7A0, 0x4E7550);
      |                                          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/mnt/storagebeast/Development/General/SilentPatch/SilentPatch/Utils/MemoryMgr.GTA.h:385:10: note: candidate: ‘T AddressByVersion(Memory::AddrVariant, Memory::AddrVariant, Memory::AddrVariant) [with T = void*; Memory::AddrVariant = std::variant<unsigned int, Memory::PatternAndOffset>]’
  385 | inline T AddressByVersion(Memory::AddrVariant address10, Memory::AddrVariant address11, Memory::AddrVariant addressSteam)
      |          ^~~~~~~~~~~~~~~~
/mnt/storagebeast/Development/General/SilentPatch/SilentPatch/Utils/MemoryMgr.GTA.h:385:47: note:   no known conversion for argument 1 from ‘int’ to ‘Memory::AddrVariant’ {aka ‘std::variant<unsigned int, Memory::PatternAndOffset>’}
  385 | inline T AddressByVersion(Memory::AddrVariant address10, Memory::AddrVariant address11, Memory::AddrVariant addressSteam)
```

(I guess I could cast all of those constants but that would add extra complexity)

File: MemoryMgr.GTA.h (hunk 2 above)
Comment (TheComputerGuy96) [reply MemoryMgr.GTA.h]: Done

File: MemoryMgr.GTA.h (hunk 1 above)
Comment (CookiePLMonster) [reply MemoryMgr.GTA.h]: That's the signed one, and I'm wondering if there's any point in keeping the unsigned variant.

Comment (TheComputerGuy96) [reply MemoryMgr.GTA.h]: Removing the unsigned variant does work with MinGW GCC (but MSVC might be different)

Comment (CookiePLMonster) [reply MemoryMgr.GTA.h]: I will verify that locally on msvc later. This functionality only exists for GTA so there is no risk of breaking other games.

File: MemoryMgr.h [hunk 3]
```diff
@@ -6,17 +6,28 @@
 
 #define WRAPPER __declspec(naked)
 #define DEPRECATED __declspec(deprecated)
+#define WRAPARG(a) ((int)a)
+
+#ifdef _MSVC_VER
```
Comment (CookiePLMonster) [inline MemoryMgr.h]: It's `_MSC_VER`.

File: MemoryMgr.h [hunk 4]
```diff
@@ -6,17 +6,28 @@
...
 #define EAXJMP(a) { _asm mov eax, a _asm jmp eax }
 #define VARJMP(a) { _asm jmp a }
-#define WRAPARG(a) ((int)a)
+#elif defined(__GNUC__) || defined(__clang__)
+#define EAXJMP(a) { __asm__ volatile("mov eax, %0\n" "jmp eax" :: "i" (a)); }
+#define VARJMP(a) { __asm__ volatile("jmp %0" :: "m" (a)); }
+#endif
 
+#ifdef _MSVC_VER
```
Comment (CookiePLMonster) [inline MemoryMgr.h]: Ditto.

**Review by CookiePLMonster — CHANGES_REQUESTED** [review]
MSVC check has to be changed, looks good otherwise.

With this one change done locally, I tested the following:
* **SilentPatch** (GTA) - v141_xp toolset, 32-bit, works fine.
* **SilentPatchCMR3** - v143 toolset, 32-bit, works fine.
* **SilentPatchYRC** (with fixes for the latest ModUtils) - v143 toolset, 64-bit, works fine.

Once this one issue is fixed, I will merge.

File: MemoryMgr.h (hunk 3 above)
Comment (TheComputerGuy96) [reply MemoryMgr.h]: That was a fairly easy definition to confuse (anyway I changed that)

File: MemoryMgr.h (hunk 4 above)
Comment (TheComputerGuy96) [reply MemoryMgr.h]: Done

**Comment by CookiePLMonster** [comment]
All good now, thank you!

EDIT:
One more question, were you not getting conversion warnings on `AddressByVersion` in III and VC that use `uintptr_t address10, uintptr_t address11, uintptr_t addressSteam` directly? I'm thinking if I should turn all those `uintptr_t`'s into `intptr_t`'s in a follow-up change. The results are being typecasted anyway and only the internals are operating on integers.

**Comment by TheComputerGuy96** [comment]
I just realized `std::basic_string<unsigned char>` (or `uint8_t`) isn't part of the C++ standard and so recent llvm-mingw can't compile the Patterns stuff that's required for all of SilentPatch (thankfully C++20 introduces a new `char8_t` type which is basically identical to those previous types but there's the last 1% missing so I have to do some type hackery)

I'll have to submit another MR to work around that (and potentially fix another issue)

**Comment by CookiePLMonster** [comment]
The lowest common denominator for ModUtils is C++17, so if possible try to make it work there. It's weird though, I thought `basic_string` can take any arbitrary type?

**Comment by TheComputerGuy96** [comment]
> I thought basic_string can take any arbitrary type?

The arbitrary type support is compiler-specific (here are the standardized types: https://en.cppreference.com/w/cpp/string/char_traits)

**Comment by CookiePLMonster** [comment]
Would defining custom traits for uint8_t and making these `basic_string` use it be an option? It could probably mostly inherit from `char_traits<char>`.

EDIT:
Seems pretty easy: https://github.com/nlohmann/json/blob/develop/include/nlohmann/detail/meta/type_traits.hpp#L194-L216

So if you go for private `pattern_traits` for those patterns, I will accept the PR.

**Pull request merged**